File: src/calcio_manager/models/competition.py
"""Competition and standings data models."""

from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from calcio_manager.models.config import CompetitionRules
from calcio_manager.models.enums import (
    AgeCategory,
    CompetitionPhase,
    CompetitionType,
    Division,
    GameFormat,
)


class StandingRow(BaseModel):
    """A single row in the league standings."""

    team_id: UUID
    team_name: str = ""
    played: int = 0
    wins: int = 0
    wins_penalties: int = 0
    losses_penalties: int = 0
    losses: int = 0
    goals_for: int = 0
    goals_against: int = 0
    points: int = 0

    # Head-to-head tracking for tiebreakers
    h2h_points: dict[str, int] = Field(
        default_factory=dict,
        description="Points in head-to-head matches: opponent_id -> points",
    )
    h2h_goals_for: dict[str, int] = Field(default_factory=dict)
    h2h_goals_against: dict[str, int] = Field(default_factory=dict)

    # Discipline
    yellow_cards: int = 0
    red_cards: int = 0

    @property
    def goal_difference(self) -> int:
        """Goal difference."""
        return self.goals_for - self.goals_against

    @property
    def matches_total(self) -> int:
        """Total matches (should equal played)."""
        return self.wins + self.wins_penalties + self.losses_penalties + self.losses

    @property
    def discipline_score(self) -> int:
        """Discipline score for tiebreaker (lower is better)."""
        return self.yellow_cards + self.red_cards * 3


class Competition(BaseModel):
    """A CSI competition (league, cup, or tournament)."""

    id: UUID = Field(default_factory=uuid4)
    name: str
    format: GameFormat = GameFormat.C7
    category: AgeCategory = AgeCategory.OPEN
    competition_type: CompetitionType = CompetitionType.LEAGUE
    phase: CompetitionPhase = CompetitionPhase.TERRITORIAL
    division: Division = Division.SERIE_ORO
    province: str = "Varese"
    region: str = "Lombardia"
    team_ids: list[UUID] = Field(default_factory=list)
    standings: list[StandingRow] = Field(default_factory=list)
    match_ids: list[UUID] = Field(default_factory=list)
    rules: CompetitionRules = CompetitionRules()
    current_match_day: int = 0
    total_match_days: int = 0
    completed: bool = False

    @property
    def num_teams(self) -> int:
        """Number of teams in the competition."""
        return len(self.team_ids)

    def get_sorted_standings(self) -> list[StandingRow]:
        """Return standings sorted by CSI tiebreaker rules.

        CSI tiebreaker order:
        1. Points
        2. Head-to-head points
        3. Head-to-head goal difference
        4. Overall goal difference
        5. Goals scored
        6. Discipline (fewer cards = better)
        """

        def sort_key(row: StandingRow) -> tuple[int, int, int, int, int, int]:
            return (
                row.points,
                0,  # H2H points computed separately for tied teams
                0,  # H2H goal diff computed separately
                row.goal_difference,
                row.goals_for,
                -row.discipline_score,  # Negative because lower is better
            )

        return sorted(self.standings, key=sort_key, reverse=True)

File: src-tauri/src/domain/models/competition.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::config::CompetitionRules;
use super::enums::{AgeCategory, CompetitionPhase, CompetitionType, Division, GameFormat};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StandingRow {
    pub team_id: Uuid,
    pub team_name: String,
    pub played: u32,
    pub wins: u32,
    pub wins_penalties: u32,
    pub losses_penalties: u32,
    pub losses: u32,
    pub goals_for: u32,
    pub goals_against: u32,
    pub points: u32,
    pub h2h_points: HashMap<String, u32>,
    pub h2h_goals_for: HashMap<String, u32>,
    pub h2h_goals_against: HashMap<String, u32>,
    pub yellow_cards: u32,
    pub red_cards: u32,
}

impl StandingRow {
    pub fn new(team_id: Uuid, team_name: impl Into<String>) -> Self {
        Self {
            team_id,
            team_name: team_name.into(),
            played: 0,
            wins: 0,
            wins_penalties: 0,
            losses_penalties: 0,
            losses: 0,
            goals_for: 0,
            goals_against: 0,
            points: 0,
            h2h_points: HashMap::new(),
            h2h_goals_for: HashMap::new(),
            h2h_goals_against: HashMap::new(),
            yellow_cards: 0,
            red_cards: 0,
        }
    }

    pub fn goal_difference(&self) -> i32 {
        self.goals_for as i32 - self.goals_against as i32
    }

    pub fn discipline_score(&self) -> u32 {
        self.yellow_cards + self.red_cards * 3
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Competition {
    pub id: Uuid,
    pub name: String,
    pub girone: String,
    pub format: GameFormat,
    pub category: AgeCategory,
    pub competition_type: CompetitionType,
    pub phase: CompetitionPhase,
    pub division: Division,
    pub province: String,
    pub region: String,
    pub team_ids: Vec<Uuid>,
    pub standings: Vec<StandingRow>,
    pub match_ids: Vec<Uuid>,
    pub rules: CompetitionRules,
    pub current_match_day: u32,
    pub total_match_days: u32,
    pub completed: bool,
}

impl Competition {
    pub fn new(
        girone: impl Into<String>,
        province: impl Into<String>,
        region: impl Into<String>,
        team_ids: Vec<Uuid>,
        rules: CompetitionRules,
    ) -> Self {
        Self {
            id: Uuid::new_v4(),
            name: String::new(),
            girone: girone.into(),
            format: GameFormat::C7,
            category: AgeCategory::Open,
            competition_type: CompetitionType::Campionato,
            phase: CompetitionPhase::Territoriale,
            division: Division::SerieOro,
            province: province.into(),
            region: region.into(),
            team_ids,
            standings: vec![],
            match_ids: vec![],
            rules,
            current_match_day: 0,
            total_match_days: 0,
            completed: false,
        }
    }

    pub fn num_teams(&self) -> usize {
        self.team_ids.len()
    }
}
